catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,396 @@
+   *     not complete sooner than the timeout when polling the {@link Source}; or {@link
+   *     Optional#empty()} if this case should be ignored
+   * @param optionalValueOnException {@link Optional} of the value to return when {@link
+   *     #getCachedValue(Source)} is called for a {@link Source} if, during the last time that
+   *     {@link #pollAllSources(long, TimeUnit, List)} was called, the {@code getValueFunction}
+   *     threw a {@link RuntimeException}; or {@link Optional#empty()} if this case should be
+   *     ignored
+   * @throws IllegalArgumentException if any of the parameters are {@code null}
    */
-  public SourcePoller(SourcePollerRunner incomingRunner) {
-
-    this.runner = incomingRunner;
-
-    scheduler =
-        Executors.newSingleThreadScheduledExecutor(
-            StandardThreadFactoryBuilder.newThreadFactory("sourcePollerThread"));
+  public SourcePoller(
+      final Function<Source, T> getValueFunction,
+      final Optional<T> optionalValueOnTimeout,
+      final Optional<T> optionalValueOnException) {
+    this.getValueFunction = notNull(getValueFunction);
+    this.optionalValueOnTimeout = notNull(optionalValueOnTimeout);
+    this.optionalValueOnException = notNull(optionalValueOnException);
+  }
 
-    handle = scheduler.scheduleAtFixedRate(runner, INITIAL_DELAY, interval, TimeUnit.MINUTES);
+  public void destroy() {
+    MoreExecutors.shutdownAndAwaitTermination(checkThreadPool, 5, TimeUnit.SECONDS);
+    MoreExecutors.shutdownAndAwaitTermination(watcherThreadPool, 5, TimeUnit.SECONDS);
+    LOGGER.debug("Destroyed SourcePoller");
   }
 
   /**
-   * Retrieves a {@link CachedSource} which contains cached values from the specified {@link
-   * Source}. Returns a {@link Source} with values from the last polling interval. If the {@link
-   * Source} is not known, null is returned.
+   * This method does not block.
    *
-   * @param source the source to get the {@link CachedSource} for
-   * @return a {@link CachedSource} which contains cached values
+   * @return the last known {@link #getCachedValue(Source)} value for the {@code source} at the last
+   *     time {@link #pollAllSources(long, TimeUnit, List)} was called, or {@link Optional#empty()}
+   *     if the {@code source} has not (yet) been polled
+   * @throws NullPointerException if the {@code source} is {@code null}
    */
-  public CachedSource getCachedSource(Source source) {
-    return runner.getCachedSource(source);
+  @SuppressWarnings(
+      "squid:S2789" /*The cache contains Optional<T> values. The Optional<T> returned by cache.getIfPresent is checked for null to see if the cache has an entry for the key*/)
+  public Optional<T> getCachedValue(final Source source) {
+    final SourceKey sourceKey = new SourceKey(source);
+    LOGGER.trace("Getting the cached value for SourceKey {}", sourceKey);
+
+    final Optional<T> cacheValue = cache.getIfPresent(sourceKey);
```

Comment:
There is already code comment about this, but basically, there isn't any real need to do it. The only thing it helps with is logging in the `getCachedValue`.
```java
final Optional<T> cacheValue = cache.getIfPresent(sourceKey);
if (cacheValue != null) {
  if (cacheValue.isPresent()) {
    final T cachedValue = cacheValue.get();
    LOGGER.trace("The cached SourceStatus for SourceKey {} is {}", sourceKey, cachedValue);
    return Optional.of(cachedValue);
  } else {
    LOGGER.debug(
        "The source for SourceKey {} was created (or modified in a way that cause its SourceKey to be different) before the last poll started, but the process to get the new value has not yet completed (or there were threading issues--see warning logs). Returning SourceStatus of \"unknown\"",
        sourceKey);
    return Optional.empty();
  }
} else {
  LOGGER.debug(
      "Unknown SourceKey {}. This only happens for unregistered sources or for sources that have been created (or modified in a way that cause its SourceKey to be different) since the last poll. Returning SourceStatus of \"unknown\"",
      sourceKey);
  return Optional.empty();
}
```
 If the cache was updated to contain T values instead, we would lose the ability in the `getCachedValue` to know is the source argument is an unknown source or if the source that hasn't finished being polled yet. The current `getCachedValue` returns the same value in both of these cases, so it might be worthwhile to forgo the logging in exchange for simpler cache code. I think I might go forward with this because it will make addressing https://github.com/codice/ddf/pull/3858#discussion_r231991791 easier too.